### DISCUSSION TITLE
Use with Apollo Server and comparison to other tools

### DISCUSSION THREAD

**Discussion by lorensr** [prisma/prisma#2076] [Q&A] [answered]
What's the recommended / possible ways to use Prisma with Apollo Server?
- delegate queries in resolvers, like old graphcool/prisma API
- use as ORM inside an Apollo data source
- use as ORM outside of a data source

How does it compare to SQLDataSource and join-monster? Does Primsa do:
- dataloader-like memoization or batching
- individual item caching
- full db query caching
- specific JOIN or SELECT based on resolver `info` / which db fields are present in the GraphQL selection set

What does Prisma do that the other things don't do? For instance:
- migrations
- studio
- non-SQL dbs (future)

**Comment by Sytten** [comment]
I will let the devs comment on this one (also interested), but I can share a bit what I have been doing with Apollo.
1. We do SDL-first development and we extend the Query with a new field
2. This field usually maps to a model in the database and we retrieve it in its resolver
3. Each field that would require a join in SQL (include in prisma), we create a resolver for it and we use the foreign key from the parent to resolve it.
4. When needed we use dataloaders to limit the number of requests to the database. Even though prisma has a dataloader/batching internally, it is still quite limited (they are working on it, you can lookup the issues). I don't believe their is other caching and I frankly prefer it that way, I don't want my database layer to be too smart.

If you read the doc, you will see that migrations are a work in progress. The studio is also experimental at this point but on the roadmap. The no-SQL DB on the roadmap too, but IMO all you probably need is postgres with a JSON column (I dislike NoSQL very much).

**Comment by nikolasburk** [comment, marked as answer]
Hey Loren, here are some brief answers to your points. I'll try to elaborate a bit later since I currently have don't have too much time but want to provide the Gist already. Also happy to jump on a call with you to discuss this further!

We also have a dedicated page about Prisma and GraphQL in the docs: https://www.prisma.io/docs/understand-prisma/prisma-in-your-stack/graphql

> What's the recommended / possible ways to use Prisma with Apollo Server?

You'd use Prisma the same way as any other DB access library (e.g. ORMs like Sequelize or SQL query builders like knex): Prisma Client is your interface to the database and is used inside of your GraphQL resolvers to send DB queries.

We have examples for what this can look like here: 

- https://github.com/prisma/prisma-examples/tree/master/typescript/graphql-apollo-server (**code-first** with Nexus schema construction library)
- https://github.com/prisma/prisma-examples/tree/master/typescript/graphql-sdl-first (**SDL-first** and uses graphql-yoga which can be easily swapped out against Apollo Server)

> delegate queries in resolvers, like old graphcool/prisma API

That's not supported with Prisma 2, see here: https://www.prisma.io/docs/more/faq#does-prisma-client-support-graphql-schema-delegation-and-graphql-binding

> use as ORM inside an Apollo data source

I'm not familiar enough with Apollo Data Sources but I believe this would be possible (but would require that someone actually builds drop-in Prisma Data Source for Apollo?).

> use as ORM outside of a data source

I guess that's the way indeed.

> How does it compare to SQLDataSource and join-monster? Does Prisma do:

I don't have much experience with these unfortunately either, but we recently employed some "data loader" optimizations in Prisma Client. Maybe @timsuchanek can speak to this a bit?

> What does Prisma do that the other things don't do? For instance:

> migrations

Yes, [Prisma Migrate](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-migrate) is available as well when using Prisma Client (since both are based on the Prisma schema). Note however that it's currently considered experimental and therefore shouldn't be used in production apps yet.

> studio

Yes, viewing and editing data in Prisma Studio is awesome! It supports a "table" and a "tree" view. The tree view is especially nice to drill into deeply nested data. You can get an overview here: https://prisma-studio.now.sh/

> non-SQL dbs (future)

Yes, in the future Prisma will support non-SQL DBs and it's also on our roadmap to figure out how to join data across different DBs!

---

One more note: Our Prisma Labs team is currently working on Nexus: https://www.nexusjs.org/#/

Nexus is a framework for building code-first  GraphQL APIs. It's not opinionated on the HTTP layer, so you can also use it together with Apollo Server.

Hope that helps for now, let me know if you have further questions!

**Comment by nikolasburk** [comment]
Some quick info from @schickling regarding the data loader questions:

- dataloader-like memoization or batching: **Yes**
- individual item caching: **Eventually**
- full db query caching: **Eventually**
- specific JOIN or SELECT based on resolver info / which db fields are present in the GraphQL selection set: **Not right now, but definitely could be built**

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
